### PR TITLE
add int_rounding_mode

### PR DESCRIPTION
**Pull request by Angryrou**
Several logistic issues in MOGD are resolved.

- [x] `best_iter` is initialized with 0. The initial configuration might be invalid if it cannot meet constraints. The current logic stops searching when no results are returned in the first `patience` steps, which is unexpected. 
    ```python
    best_iter: Optional[int] = None
    ...
    if best_iter is not None and i > best_iter + self.patience:
        break
    ``` 

- [x] We currently recorded the best loss after input variables trigger `backward`, which does not alligh with the variables corresponding to the corresponding loss in the unprocessed scenario.

- [x] Rounding overhead is high. The current design requires 1-2 model forward paths per tuning iteration, which is heavier than expected. After a second thought, I provided two methods that need exactly one forward path per tuning iteration, controlled by a bool variable `strict_rounding`
    - when `strict_rounding = False`, we round int variables only once at the end of all iterations.
    - when `strict_rounding = True`, we round int variables at the end of each iteration. (this actually avoids recomputing the objective bc the input variable for each iteration has already been rounded at the end of the previous iteration) 

- [x] Unprocessed input does not handle `input_parameters` properly.

### CODE REVIEW

**Comment by Angryrou** [comment]
Current dependencies on/for this PR:
* `dev`
  * **PR #6** <a href="https://app.graphite.dev/github/pr/Angryrou/udao/6?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈
    * **PR #9** <a href="https://app.graphite.dev/github/pr/Angryrou/udao/9?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **PR #10** <a href="https://app.graphite.dev/github/pr/Angryrou/udao/10?utm_source=stack-comment-icon" target="_blank">Graphite</a> 

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/Angryrou/udao/6?utm_source=stack-comment).

File: udao/optimization/soo/mogd.py [hunk 1]
```diff
@@ -751,19 +771,39 @@ def objective_loss(
...
             constraint_loss = self.constraints_loss(const_outputs, problem.constraints)
 
         loss = objective_loss + constraint_loss
-        return loss
+        sum_loss = th.sum(loss)
+        min_loss = th.min(loss).cpu().item()
+        min_loss_id = int(th.argmin(loss).cpu().item())
+        best_obj = obj_output[min_loss_id].cpu().item()
+
+        return (
```
Comment (julhien) [inline udao/optimization/soo/mogd.py]: It's better to either return fewer elements (splitting into smaller functions) or return a dictionary to keep clarity on what's returned.
A long tuple can lead to difficulty reading and mistakes in the order of returned values

File: udao/optimization/soo/mogd.py [hunk 2]
```diff
@@ -426,111 +455,102 @@ def _processed_single_start_opt(
...
+        if best_iter is not None:
+            assert best_obj is not None and best_feature_input is not None
+            with th.no_grad():
+                best_feature_input = cast(th.Tensor, best_feature_input)
+                feature_container = make_tabular_container(best_feature_input)
+                best_raw_df = problem.data_processor.inverse_transform(
+                    feature_container, "tabular_features"
+                )
+                if not self.strict_rounding:
+                    best_raw_vars: Dict[str, Any] = {
```
Comment (julhien) [inline udao/optimization/soo/mogd.py]: In both processed and unprocessed case, rounding could be factored into a rounding function since if is repeated twice in the code.

File: udao/optimization/soo/mogd.py [hunk 3]
```diff
@@ -751,19 +771,39 @@ def objective_loss(
...
 
-    def compute_loss(self, problem: co.SOProblem, input_data: Any) -> th.Tensor:
-        obj_output = problem.objective.function(input_data)
+    def _compute_loss(
+        self, problem: co.SOProblem, input_data: Union[UdaoInput, Dict]
+    ) -> Tuple[th.Tensor, float, float, int, bool]:
+        obj_output = (
+            problem.objective.function(input_data)  # type: ignore
+            if isinstance(input_data, UdaoInput)
+            else problem.objective.function(**input_data)
```
Comment (julhien) [inline udao/optimization/soo/mogd.py:782]: Sorry I missed this. 
I like the fact that the lower level function makes the distinction between the two kinds of inputs. 

If we have time (other PR) it would make the logic clearer to have a single optimization loop that calls different specialized functions that can make a difference between different kinds of inputs.

File: udao/optimization/tests/soo/test_mogd.py [hunk 4]
```diff
@@ -159,21 +163,32 @@ def test_solve(
...
-            (co.FloatVariable(0, 24), 148.17662, {"cores": 16.2}, 1),
-            (co.FloatVariable(0, 24), 148.17662, {"cores": 16.2}, 16),
-            (co.IntegerVariable(0, 24), 150, {"cores": 16}, 1),
-            (co.IntegerVariable(0, 24), 150, {"cores": 16}, 16),
+            (
+                True,
+                co.IntegerVariable(0, 24),
+                1,
+                200,
+                {"cores": 12},
```
Comment (julhien) [inline udao/optimization/tests/soo/test_mogd.py]: Do you know why optimization "fails" in this case?

**Review by julhien — CHANGES_REQUESTED** [review]

File: udao/optimization/soo/mogd.py (hunk 2 above)
Comment (Angryrou) [reply udao/optimization/soo/mogd.py]: postpone this to a one-loop PR in the future

File: udao/optimization/tests/soo/test_mogd.py (hunk 4 above)
Comment (Angryrou) [reply udao/optimization/tests/soo/test_mogd.py]: The rounding at the end of iteration can cancel out the gradient descent in some unluck cases

**Review by julhien — APPROVED** [review]
